### PR TITLE
Upgrading cosmosdb statestore to track2SDK

### PR DESCRIPTION
**Pull request by RyanLettieri**
Signed-off-by: Ryan Lettieri <ryanLettieri@microsoft.com>

# Description

Updating cosmosdb state to use track2 sdk

### CODE REVIEW

**Comment by berndverst** [comment]
@ItalyPaleAle can you do the first review of this?

**Comment by berndverst** [comment]
Let's merge this AFTER https://github.com/dapr/components-contrib/pull/1972

This way we can make sure that the SDK upgrade didn't break the certification test.

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:17]: Delete this comment and remove the `_ "embed"` import as those are not necessary anymore. Also, you should delete everything in the `storedprocedures` folder

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:87]: Misses a duration. Is this in seconds? SHould be `20 * time.Second` then

I would also consider calling this something like `defaultTimeout`

Comment (ItalyPaleAle) [reply state/azure/cosmosdb/cosmosdb.go:87]: Also, nit, but plese put all constants in the same block (see lines 79-84 above)

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:90]: Return type must be an interface now, so `state.Store`

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:135]:
```go
var cred azcosmos.KeyCredential
cred, err = azcosmos.NewKeyCredential(...)
```

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:145]: Same here, and on line 149. Would be best to avoid defining new kinds of error variables.

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:92]: Is there a reason why the features need to be stored as a property in the object? They could just be returned by the `Features()` method without being a property in the object.

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:271]: Please make sure to add a timeout (using the defaultTimeout constant) to the context

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:343]: Should probably check if not empty too:

```
if req.ETag != nil && *req.ETag != ""
```

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:357]: Should probably check if not empty too:

```
if req.ETag != nil && *req.ETag != ""
```

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [reply state/azure/cosmosdb/cosmosdb.go:375]: Perhaps this should be configurable via a metadata option, since there can be many operations here?

File: state/azure/cosmosdb/cosmosdb_query.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb_query.go:34]: Nit: newline after this please

File: state/azure/cosmosdb/cosmosdb_query.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb_query.go:162]: Timeout here too please

File: state/azure/cosmosdb/cosmosdb_query.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb_query.go:177]: did you mean `&items[i].Etag`?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

**Comment by berndverst** [comment]
Conformance: https://github.com/dapr/components-contrib/actions/runs/3018674281
Certification: https://github.com/dapr/components-contrib/actions/runs/3018674217

**Comment by berndverst** [comment]
@RyanLettieri please make sure that the conformance test for CosmosDB State store passes in this run: https://github.com/dapr/components-contrib/actions/runs/3019169073

You can manually launch another run via `/ok-to-test` in a new comment.

**Review by ealsur — DISMISSED** [review]

**Comment by berndverst** [comment]
@RyanLettieri converting this to draft until you get all tests to pass locally

**Comment by RyanLettieri** [comment]
Cert: https://github.com/dapr/components-contrib/actions/runs/3063891151

Conformance: https://github.com/dapr/components-contrib/actions/runs/3063891192

File: state/azure/cosmosdb/cosmosdb.go
Comment (berndverst) [inline state/azure/cosmosdb/cosmosdb.go:353]: You should use `populatePartitionMetadata(key, request.Metadata)`

That way it will use the item key if the `partitionKey` isn't explicitly defined.

File: state/azure/cosmosdb/cosmosdb.go
Comment (berndverst) [reply state/azure/cosmosdb/cosmosdb.go:403]: @RyanLettieri please address Alessandro's comment.

**Review by berndverst — CHANGES_REQUESTED** [review]
Left a comment about the partition key that needs to be addressed.

And please look at Alessandro's comment about the timeout.

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:428]: Should this return an error instead?

File: state/azure/cosmosdb/cosmosdb.go
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:426]: Why this here?

Comment (ItalyPaleAle) [reply state/azure/cosmosdb/cosmosdb.go:426]: I understand that this was here from before, but before this was in a retry block... Do we need this check now? What if the error is a 429, what should we do?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

**Review by ItalyPaleAle — DISMISSED** [review]

**Comment by artursouza** [comment]
I propose this component comes out as v2 and we keep the old one as v1. Then, we update the tests to run against both. I don't think our certification tests for state store has validations cross versions - how data written from an older version can be read from the new version - so, given a significant change like this, there would be less risk to upgrade to a new version. I know it is a subjective call but I am concerned about the cross-version compatibility with such change refactoring (or rewrite).

/cc @yaron2 thoughts?

**Comment by yaron2** [comment]
> I propose this component comes out as v2 and we keep the old one as v1. Then, we update the tests to run against both. I don't think our certification tests for state store has validations cross versions - how data written from an older version can be read from the new version - so, given a significant change like this, there would be less risk to upgrade to a new Dapr release if customers can opt in to use the new coponent. I know it is a subjective call but I am concerned about the cross-version compatibility with such change as this rewrite.
> 
> /cc @yaron2 thoughts?

As discussed, I agree. I am concernrd the new SDK might carry bugs and changes that are not cought by our tests.

**Comment by ItalyPaleAle** [comment]
The concern I have with keeping the old "v1" around is that we need to maintain the dependency on the a8m SDK, which is sort-of unmaintained right now and I'd love to get rid of that for good.

Is there a way we could perhaps keep both implementations in 1.9, making the new one an "opt in" via a feature flag, and once we feel comfortable (say, 1.10), switch to the new one only?

**Comment by RyanLettieri** [comment]
Cert: https://github.com/dapr/components-contrib/actions/runs/3092835163
Conf: https://github.com/dapr/components-contrib/actions/runs/3092835186

**Review by berndverst — APPROVED** [review]

**Comment by berndverst** [comment]
In the maintainers chat we agreed we can keep the current v1 designation and don't need to introduce a new version in this case.

**Comment by berndverst** [comment]
Unfortunately this PR removed the user agent so that we lost our ability to track users of Dapr + CosmosDB.

**Comment by ealsur** [comment]
@ItalyPaleAle The TelemetryOptions are common to all Azure SDKs for Go, yeah

**Comment by berndverst** [comment]
I already opened a PR for it. And yes we have done this.

Please do not discuss further here.

**Pull request merged**
